rust-src/concordium_base/src/curve_arithmetic/ed25519_instance.rs

```diff
@@ -0,0 +1,375 @@
+
+    fn sub_assign(&mut self, other: &Self) { self.0.sub_assign(other.0) }
+
+    fn mul_assign(&mut self, other: &Self) { self.0.mul_assign(other.0) }
+
+    fn inverse(&self) -> Option<Self> {
+        if self.is_zero() {
+            None
+        } else {
+            Some(self.0.invert().into())
+        }
+    }
+}
+
+impl PrimeField for RistrettoScalar {
+    // Taken from `curve25519-dalek` v.4.1.1 that implements `ff::PrimeField`
+    const CAPACITY: u32 = 252;
+    // Taken from `curve25519-dalek` v.4.1.1 that implements `ff::PrimeField``
+    const NUM_BITS: u32 = 253;
+
+    fn into_repr(self) -> Vec<u64> {
+        let mut vec: Vec<u64> = Vec::new();
+        let bytes: [u8; 32] = self.0.to_bytes();
+        for chunk in bytes.chunks_exact(8) {
+            // The chunk size is always 8 and there is no remainder after chunking, since
+            // the representation is a 32-byte array. That is why it is safe to
+            // unwrap here.
+            let x: [u8; 8] = chunk.try_into().unwrap();
+            let x_64 = u64::from_le_bytes(x);
+            vec.push(x_64);
+        }
+        vec
+    }
+
+    fn from_repr(r: &[u64]) -> Result<Self, super::CurveDecodingError> {
+        let tmp: [u64; 4] = r
+            .try_into()
+            .map_err(|_| super::CurveDecodingError::NotInField(format!("{:?}", r)))?;
+        let mut s_bytes = [0u8; 32];
+        let mut offset = 0;
+        for x in tmp {
+            let max = offset + 8;
+            LittleEndian::write_u64(&mut s_bytes[offset..max], x);
+            offset = max;
+        }
+        let res = Scalar::from_canonical_bytes(s_bytes).ok_or(
+            super::CurveDecodingError::NotInField(format!("{:?}", s_bytes)),
+        )?;
+        Ok(res.into())
+    }
+}
+
+impl Serial for RistrettoPoint {
+    fn serial<B: Buffer>(&self, out: &mut B) {
+        let compressed_point = self.compress();
+        let res: &[u8; 32] = compressed_point.as_bytes();
+        out.write_all(res)
+            .expect("Writing to a buffer should not fail.");
+    }
+}
+
+impl Deserial for RistrettoPoint {
+    fn deserial<R: byteorder::ReadBytesExt>(source: &mut R) -> crate::common::ParseResult<Self> {
+        let mut buf: [u8; 32] = [0; 32];
+        source.read_exact(&mut buf)?;
+        let res = CompressedRistretto::from_slice(&buf)
+            .decompress()
+            .ok_or(anyhow::anyhow!("Failed!"))?;
+        Ok(res)
+    }
+}
+
+impl Curve for RistrettoPoint {
+    type MultiExpType = RistrettoMultiExpNoPrecompute;
+    type Scalar = RistrettoScalar;
+
+    const GROUP_ELEMENT_LENGTH: usize = 32;
+    const SCALAR_LENGTH: usize = 32;
+
+    fn zero_point() -> Self { Self::identity() }
+
+    fn one_point() -> Self { RISTRETTO_BASEPOINT_POINT }
+
+    fn is_zero_point(&self) -> bool { self == &Self::zero_point() }
+
+    fn inverse_point(&self) -> Self { -self }
+
+    fn double_point(&self) -> Self { self + self }
```

Comment:
I know there was discussion about this, but there is no more efficient method exposed than this?

If there isn't at the moment, perhaps add a comment that the reason for implementing it like this is that there is no more efficient method.

Reply:
I could not find a direct implementation of  `double` in the dalek crate. There is [double_and_compress_batch](https://docs.rs/curve25519-dalek/latest/curve25519_dalek/ristretto/struct.RistrettoPoint.html#method.double_and_compress_batch), but this does not look as a good option, since it produces a compressed representation that requires decompressing back. I tried to make it work, but somehow the calls failed on decompressing after doubling, so I did not investigate it further.

Moreover, newer versions implement double [using addition ](https://docs.rs/curve25519-dalek/4.1.1/src/curve25519_dalek/ristretto.rs.html#1203-1205). For `EdwardsPoint` (which `RistrettoPoint` wraps), there is a doubling operation, but it seems to be not available through the Ristretto representation.

In general, it does not seem to be a problem, the performance of in range proofs was the same as using the [bulletproofs](https://crates.io/crates/bulletproofs) crate (our was even sligthly faster).

I'll leave a comment in the code pointing this out.